lms/envs/common.py

```diff
@@ -1350,6 +1350,8 @@ def _make_locale_paths(settings):
 # Platform for Privacy Preferences header
 P3P_HEADER = 'CP="Open EdX does not have a P3P policy."'
 
+MEMORY_GRAPH_DIRECTORY = os.path.join(tempfile.mkdtemp(prefix='memory_graphs'), 'lms_{}'.format(os.getpid()))
```

Comment:
>We restart gunicorn processes after some number of requests on prod (I can never find that number when I look for it in our config repos).

For future reference:

https://github.com/edx/edx-internal/blob/617c3762ea/ansible/vars/prod-edx.yml#L42
```
EDXAPP_LMS_MAX_REQ: 2000
EDXAPP_CMS_MAX_REQ: 100
```

>Given the traffic we normally see, how many of these directories will we get if the machine stays up for a week? Do we have to worry that having all those dirs will break something?

We really don't intend on memory graphs being ever collected in prod, let alone in loadtest for more than a few minutes.  Perhaps those intentions should be clarified in the middleware docstring.

Reply:
Right, but this is in common.py, so these temp directories are going to be created on every start in all environments where it's not overridden, which includes our production environments, right?

Reply:
Oh, right.  We didn't used to call tempfile.mkdtemp in settings before, and the `lms_<pid>` dir is just created on-demand.  With this code, I think the following hierarchy would result:

```
/tmp/memory_graphs1111/lms_2222
/tmp/memory_graphs3333/lms_4444
...
```

which is a little bizarre since we normally expect the following in devstack:

```
memory_graphs/lms_1111
memory_graphs/lms_2222
```

I think we should push down the mkdtemp call to the point in code where the graphs are generated and the final directory is created on-demand.

Reply:
Given our current settings (which are in edx-internal), we'd be creating about 700 of these per day per lms worker and 300 on cms. The ext2/ext3 subdirectory limit is 30,000, higher for ext4 and such. So not necessarily a problem, but not ideal either (especially considering we only use them when actively diagnosing a memory leak). I'll see if I can make this a lazy evaluation so it doesn't get created unless we're about to actually use it.

Reply:
@jmbowman why does the memory_graphs directory itself even need the uuid?  Can't we do something like this:

```
MEMORY_GRAPH_DIRECTORY = os.path.join('/tmp/memory_graphs', tempfile.mkdtemp(prefix='lms_{}_'.format(os.getpid())))
```

Reply:
In addition to lazy evaluation, can we also fix the heirarchy by appending the uuid to the subdirectory instead of the top level memory_graphs directory?

something like this, but lazy:

```
MEMORY_GRAPH_DIRECTORY = tempfile.mkdtemp(prefix='lms_{}_'.format(os.getpid()), dir='/tmp/memory_graphs')
```

Reply:
I'm a little hesitant to assume that any particular directory entry under `/tmp` isn't already in use by something else; and if we went with that last suggestion, we'd first have to ensure that the `memory_graphs` directory already existed.  Given that PIDs can eventually be reused, I actually like that `mkdtemp` guarantees us a distinct directory for each gunicorn worker.  The directory name prefix should make it easy enough to grab all the relevant data from a VM.

Reply:
For the record, I am fine with any solution that:

1. Makes it clear what pid a particular dump comes from.
2. Makes it clear when a dump was created, in case we have to match it to log data or sequence it for looking at growth/trends.
3. Doesn't make a dangerously large number of directories if left unattended for a couple of weeks.